docs/rules/no-reject-literal.md

```diff
@@ -0,0 +1,57 @@
+# disallow using literals as Promise rejection reasons (no-reject-literal)
+
+It is considered good practice to only instances of the built-in `Error` object for user-defined errors in Promises. `Error` objects automatically store a stack trace, which can be used to debug an error by determining where it came from. If a Promise is rejected with a non-`Error` value, it can be difficult to determine where the rejection occurred.
+
+
+## Rule Details
+
+This rule aims to ensure that Promises are only rejected with `Error` objects.
+
+Examples of **incorrect** code for this rule:
+
+```js
+/*eslint no-reject-literal: "error"*/
+
+Promise.reject("something bad happened");
+
+Promise.reject(5);
+
+Promise.reject();
```

Comment:
I feel like we've gotten a bit off-topic. I'd be glad to keep discussing the different things that Promises can be used for and their synchronous parallels, but maybe we should do that e.g. in a chatroom so that we don't spam peoples' inboxes.

At the moment, I'm still leaning towards :-1: for adding an exception for `undefined` as Promise rejection reasons, but I'm fine with waiting to see what the rest of the team thinks about it. (Alternatively, maybe it would be better to open a separate issue after this rule lands on master, since it would be a backwards-compatible change.)